lib/async_hooks.js

```diff
@@ -130,6 +131,60 @@ function createHook(fns) {
   return new AsyncHook(fns);
 }
 
+// AsyncLocal API //
+
+const locals = new Set();
+const localsHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const execRes = executionAsyncResource();
+    locals.forEach((local) => {
+      local[kPropagateSymbol](execRes, resource);
+    });
+  }
+});
+
+const kResToValSymbol = Symbol('resToVal');
+const kPropagateSymbol = Symbol('propagate');
+
+class AsyncLocal {
+  constructor() {
+    this[kResToValSymbol] = new WeakMap();
+    locals.add(this);
+    localsHook.enable();
+  }
+
+  [kPropagateSymbol](execRes, initRes) {
+    const value = this[kResToValSymbol].get(execRes);
+    if (value !== undefined) {
+      this[kResToValSymbol].set(initRes, value);
+    }
+  }
+
+  get() {
+    if (this[kResToValSymbol]) {
+      return this[kResToValSymbol].get(executionAsyncResource());
+    }
+    return undefined;
+  }
+
+  set(value) {
+    if (!this[kResToValSymbol]) {
+      throw new ERR_ASYNC_LOCAL_CANNOT_SET_VALUE();
+    }
+    this[kResToValSymbol].set(executionAsyncResource(), value);
+  }
+
+  remove() {
+    if (this[kResToValSymbol]) {
+      delete this[kResToValSymbol];
+      locals.delete(this);
```

Comment:
I don't really like the manual cleanup required here. If someone decides to make a unique `AsyncLocal` object per-request it may be ambiguous where to close the object and they might even not close it at all, leading to a memory leak. It'd be better to use a WeakSet so you don't have to worry about explicit cleanup.

Reply:
`WeakSet` won't do here, as it doesn't allow iteration over its items. A `Set` with `WeakRef`s + `FinalizationGroup` API could potentially help to achieve automatic `AsyncLocal` clean up, but FG API is experimental and has to be enabled via a flag.

On the other hand, other CLS API implementations do not provide a way to free underlying resources and disable hooks. That's because most applications use a single instance of CLS without getting rid of it. So, `AsyncLocal.remove()` method is an advanced API, useful in specific scenarios.

WDYT?

Reply:
There's ways to eliminate the global `Set` entirely by making the `WeakMap` global and propagating on that directly.

Reply:
A global `WeakMap` used to store values won't allow to isolate different `AsyncLocal`s, which kills the whole idea.

If you mean a `WeakMap` that stores `<AsyncLocal, WeakMap>` pairs, then it's not possible to iterate over `WeakMap` entries.

Reply:
I mean something like `WeakMap<AsyncResource, Set<AsyncLocal>>`. The hooks can loop over the set of AsyncLocal instances at the current level and propagate them downwards. It means an array at each level, rather than just the top-level, but it also cleans up automatically.

Reply:
That's an interesting idea. Another level of indirection with a `WeakMap<AsyncResource, Set<AsyncLocal>>` will help to clean up both values and `AsyncLocal`s list in the hook. As downsides I can see the following:
* The hook will have to be kept always enabled, once the first `AsyncLocal` instance was created. Currently `.remove()` disables the hook if there are no active `AsyncLocal`s.
* This change will probably have a certain performance impact. But that has to be measured.

In general, it sounds like a nice idea for an experiment. I'm going to do that experiment, if this PR gets some chances to be reviewed and merged (I understand that chances are close to zero, especially considering https://github.com/nodejs/node/pull/30959#issuecomment-568745337).